### DISCUSSION TITLE
API Diskussion

### DISCUSSION THREAD

**Discussion by darth-hp** [Secret-Lab-Productions/heizoel24-mex#2] [General]
Zur Weiterführung der Diskussion von https://github.com/ioBroker/AdapterRequests/issues/637

**Comment by cpatscheider** [comment]
So könnte das aussehen:
```
from flask import Flask
from flask import jsonify
import requests
import time
import json
import logging

app = Flask(__name__)

def login():
    app.logger.info('Login in...')
    global session_id
    global logged_in
    url = "https://api.heizoel24.de/app/api/app/Login"
    newHeaders = {'Content-type': 'application/json'}
    reply = requests.post(url, json = { "Password" : "redacted", "Username" : "redacted"}, headers=newHeaders)
    app.logger.debug('Heizoel Login Status Code: ' + str(reply.status_code))
    if reply.status_code == 200:
        app.logger.debug(reply.text)
        reply_json = json.loads(reply.text)
        if reply_json['ResultCode'] == 0:
            session_id = reply_json['SessionId']
            app.logger.info('Session ID: ' + session_id)
            logged_in = True

def get_sensor_data():  
    app.logger.info('Getting sensor data...')
    global sensor_data  
    global sensor_data_age
    global sensor_data_cache_time
    now = time.time()    
    if now - sensor_data_age > sensor_data_cache_time: # Every 4 hours
        login()
        app.logger.info('Refresh sensor data cache...')
        url = 'https://api.heizoel24.de/app/api/app/GetDashboardData/'+ session_id + '/1/1/False'
        reply = requests.get(url)     
        app.logger.info('Heizoel GetDashboardData Status Code: ' + str(reply.status_code))
        if reply.status_code == 200:   
            sensor_data_age = time.time()
            # Make the 0th element a list
            sensor_data = '[' + reply.text+ ']'
        else:
            # Return what we got from Heizoel
            return reply.text, reply.status_code
    
    app.logger.debug(sensor_data)
    return sensor_data
    

@app.route('/mex/', methods=['GET'])
def mex():  
    return get_sensor_data()

def main():
    global sensor_data_age
    global sensor_data_cache_time
    sensor_data_cache_time = 14400
    sensor_data_age = time.time() - sensor_data_cache_time
    app.logger.setLevel(logging.DEBUG)
    app.run(host='0.0.0.0', port=105)

if __name__ == '__main__':   
    main()
```

Die Antwort von Mex sieht ungefähr so aus:  
```
{
    "Items": [
        {
            "SensorId": redacted,
            "IsMain": true,
            "CurrentVolumePercentage": 65,
            "CurrentVolume": 2341,
            "NotifyAtLowLevel": 40,
            "NotifyAtAlmostEmptyLevel": 20,
            "NotificationsEnabled": false,
            "Usage": 1.228370243503071,
            "RemainsUntil": "2023-06-25T15:16:13.0277244+02:00",
            "MaxVolume": 3600,
            "ZipCode": "redacted",
            "MexName": "redacted",
            "LastMeasurementTimeStamp": "2022-07-22T03:32:02",
            "LastMeasurementWithDifferentValue": "2022-07-18T07:46:01",
            "BatteryPercentage": 75,
            "Battery": 2.9,
            "LitresPerCentimeter": 24,
            "LastMeasurementWasSuccessfully": true,
            "SensorTypeId": 1,
            "HasMeasurements": true,
            "MeasuredDaysCount": 188,
            "LastMeasurementWasTooHigh": false,
            "YearlyOilUsage": null,
            "RemainingDays": 338,
            "RemainsUntilCombined": {
                "MonthAndYear": "Juni 2023",
                "RemainsValue": "11",
                "RemainsUnit": "M"
            },
            "LastOrderPrice": 70.448000000000008,
            "ResultCode": 0,
            "ResultMessage": null
        }
    ],
    "LastOrderPrice": 0,
    "PriceComparedToYesterdayPercentage": 0,
    "PriceForecastPercentage": 1,
    "HasMultipleMexDevices": false,
    "DashboardViewMode": 1,
    "ShowComparedToYesterday": true,
    "ShowForecast": false,
    "ResultCode": 0,
    "ResultMessage": null
}
```

**Reply by ltspicer** [reply]
Hast Du selber einen MEX? Wenn nicht, mit welchen Daten könnte ich dienen, ohne gleich sensible Daten zu veröffentlichen?

**Reply by cpatscheider** [reply]
Ja ich hab einen. Man müsste das ganze mal in den Code alexiosc mergen und passwort und username als config definieren. Dann könntest du beta testen :)

**Comment by darth-hp** [comment]
@cpatscheider das funktioniert. Ich hab' lediglich meine PLZ geändert:
```json
[
    {
        "Items": [
            {
                "SensorId": 9422,
                "IsMain": true,
                "CurrentVolumePercentage": 16,
                "CurrentVolume": 597,
                "NotifyAtLowLevel": 20,
                "NotifyAtAlmostEmptyLevel": 10,
                "NotificationsEnabled": true,
                "Usage": 4.0,
                "RemainsUntil": "2024-11-13T08:16:56.5999476+01:00",
                "MaxVolume": 3800,
                "ZipCode": "12345",
                "MexName": "Heizölmessgerät",
                "LastMeasurementTimeStamp": "2024-02-09T02:05:02",
                "LastMeasurementWithDifferentValue": "2024-02-05T13:05:02",
                "BatteryPercentage": 100,
                "Battery": 3.0,
                "LitresPerCentimeter": 29,
                "LastMeasurementWasSuccessfully": true,
                "SensorTypeId": 1,
                "HasMeasurements": true,
                "MeasuredDaysCount": 594,
                "LastMeasurementWasTooHigh": false,
                "YearlyOilUsage": null,
                "RemainingDays": 278,
                "RemainsUntilCombined": {
                    "MonthAndYear": "November 2024",
                    "RemainsValue": "9",
                    "RemainsUnit": "M"
                },
                "LastOrderPrice": null,
                "ResultCode": 0,
                "ResultMessage": null
            }
        ],
        "LastOrderPrice": 117.95,
        "PriceComparedToYesterdayPercentage": 0.83,
        "PriceForecastPercentage": -0.13,
        "HasMultipleMexDevices": false,
        "DashboardViewMode": 1,
        "ShowComparedToYesterday": true,
        "ShowForecast": true,
        "ResultCode": 0,
        "ResultMessage": null
    }
]
```

**Reply by ltspicer** [reply]
Aber UUID benötigt man auch. Oder?

**Reply by darth-hp** [reply]
Nein - es wird automatisch eine Session ID beim Login generiert. Lediglich User/PW ist notwendig

**Comment by cpatscheider** [comment]
Ich hab mal ein paar Änderungen gemacht am Code. Mir fehlt jetzt aber das Wissen wie man das ganze testen kann.
Könnt ihr mir da helfen?
Edit: Habs mir in meine persönliche HA Instanz geworfen...

**Reply by elsi06** [reply]
Läufts?

**Reply by ltspicer** [reply]
Guckst Du hier: https://github.com/ltspicer/iobroker.mex
Ist soweit erledigt ^^

**Comment by ltspicer** [comment]
Hier mal mein Python Script, welches auf irgend einem Computer laufen kann.
Werde dann noch MQTT für den ioBroker usw integrieren.
Ist ein erster, laufender Entwurf und kann/muss noch verbessert werden.
Auch die Implementierung als sauberer Adapter fehlt natürlich noch.
Das Script kann dann per Cron alle 2 bis 4 Stunden mal aufgerufen werden.
Ich muss jetzt aber erst mal ins Bett ;)

...
Schaut hier: https://github.com/ltspicer/iobroker.mex

**Reply by ltspicer** [reply]
Verfügbare Daten:


'SensorId': 3018,
'IsMain': True,
'CurrentVolumePercentage': 73,
'CurrentVolume': 2865,
'NotifyAtLowLevel': 20,
'NotifyAtAlmostEmptyLevel': 10,
'NotificationsEnabled': False,
'Usage': 6.230769230769231,
'RemainsUntil': '2025-12-27T23:38:39.9003594+01:00',
'MaxVolume': 3920,
'ZipCode': '8888',
'MexName': 'AAAstrasse, Peter Meier',
'LastMeasurementTimeStamp': '2024-02-14T02:11:02',
'LastMeasurementWithDifferentValue': '2024-02-11T01:11:02',
'BatteryPercentage': 50,
'Battery': 2.8,
'LitresPerCentimeter': 27,
'LastMeasurementWasSuccessfully': True,
'SensorTypeId': 1,
'HasMeasurements': True,
'MeasuredDaysCount': 1029,
'LastMeasurementWasTooHigh': False,
'YearlyOilUsage': None,
'RemainingDays': 682,
'RemainsUntilCombined': {'MonthAndYear': 'Dezember 2025','RemainsValue': 'ca. 2','RemainsUnit': 'J'},
'LastOrderPrice': None,
'ResultCode': 0,
'ResultMessage': None

**Comment by ltspicer** [comment]
Hier mal die verfeinerte Version ;)

...

Schaut hier: https://github.com/ltspicer/iobroker.mex

**Comment by cpatscheider** [comment]
Ich bin für home assistant so weit, dass man das Plugin konfigurieren kann und ich im log sehe dass es sich erfolgreich einloggt. Dann aber noch nicht die Werte liest.. Würde den stand gerne mal auf einen Branch hier committen. @hphde Kannst/musst du mir hierfür eine Berechtigung erteilen?

**Comment by darth-hp** [comment]
Pull request...

**Comment by cpatscheider** [comment]
Du könntest mich auch einfach als Collaborator hinzufügen, dann müsst ich jetzt nicht wieder das repo clonen und rebasen ;)

**Comment by darth-hp** [comment]
Hab dich für Triage eingetragen

**Comment by cpatscheider** [comment]
Hab irgendwie immer noch keine Schreibrechte. Hab es jetzt mal hier geforkt: https://github.com/cpatscheider/heizoel24-mex und meinen aktuellen Stand gepusht. 
Die dortige Version lässt sich in HA konfigurieren und man sieht im Log das sie sich einloggt. Darüber hinaus bin ich nicht gekommen. Soll ich einen PR machen oder willst du dir die Version da mal ansehen?

**Comment by darth-hp** [comment]
Der grundsätzliche Submit Flow geht für mich immer über einen PR auf main, damit mindestens ein weiteres Augenpaar drauf geschaut hat. Vor allem wenn man kein Test-Flow hat. Oder bist du anderer Meinung? Typischerweise via Branch aus dem gleichen Repo oder über PR von einem anderen.
Sollte ersteres nicht möglich sein, habe ich evtl die Settings verkehrt und muss das nochmal checken

**Reply by cpatscheider** [reply]
Sehe ich genauso. Deswesgen wollte ich ja erst in dem gleichem Repo einen neuen Branch machen. Den zu erstellen und pushen durfte ich aber nicht...

**Comment by darth-hp** [comment]
Geht's jetzt?

**Comment by cpatscheider** [comment]
jup :) Danke! Feedback und Hilfe ist gerne gesehen

**Comment by darth-hp** [comment]
Mein Großprojekt ist gerade Dachausbau 😆

**Comment by ltspicer** [comment]
Mal ne scheue Frage.
Warum diskutiert ihr hier über HA implementierung? Für HA ist ja schon lange eine Lösung vorhanden.
Ich dachte, es geht hier um den ioBroker.

Bin übrigens daran, einen IOB Adapter zu erstellen. Das Python Script läuft bereits stabil.
Nur, damit wir nicht parallel daran arbeiten ^^

**Reply by darth-hp** [reply]
> Warum diskutiert ihr hier über HA implementierung? Für HA ist ja schon lange eine Lösung vorhanden.
> Ich dachte, es geht hier um den ioBroker.

Wo gibt's die @ltspicer?

**Reply by ltspicer** [reply]
Ah sorry, ist  ja Ecofrog.

**Comment by ltspicer** [comment]
> Mein Großprojekt ist gerade Dachausbau 😆

Habe 2023 eine Dachhälfte neu machen lassen und gleich eine 10kWp PV Anlage drauf getan.
Wünsche Dir viel Erfolg.

**Comment by axute** [comment]
Wie sieht es hier aus?
Funktioniert es in HA? Werde nicht ganz schlau daraus.

**Reply by Marcwa19197** [reply]
Ich habe es via MQTT angebunden. Klappt wunderbar. Eine "nativere" HA Integration wäre natürlich mega gut. 👍  Jedenfalls an der Stelle schon einmal ein dickes Danke an @ltspicer! 

Ich hatte mich auch mal rangewagt und hatte sogar die JSON Daten zwischen dem MEXX und dem entsprechenden API Server entschlüsseln können. Wollte dann etwas MITM mäßiges "bauen" um die Daten bevor sie in die Cloud gehen abzufangen. Die Lösung mit dem fetchen des "offiziellen" API-Servers ist natürlich auch genial, auch wenn es gegen meine "local only" Policy verstößt ;)

Grüße!

**Reply by ltspicer** [reply]
Hier die aktuelle Version:
https://github.com/ltspicer/heizoel24.mex

**Reply by Marcwa19197** [reply]
Jetzt habe ich schon meine MQTT-Sensoren "hart" in der HA Config... habe gesehen dass sich das MQTT-Topic ja ändert in die ID des Sensors... muss ich mal ran an die Config. Aber mega gut!

Nur mal als Crosslink... https://community.home-assistant.io/t/proteus-sensor-api-integration/463144/25

**Reply by Marcwa19197** [reply]
Wäre den eine "Local only" Lösung eine Option? Soll ich zumindest mal einen neuen Diskussionsthread öffnen?
Ich hätte da nämlich auch ein Tool um die Konfig des Mex abzuändern. Dort kann man dann auch den API Server eintragen etc....  Liegt irgendwo auf einer Windoof Büchse rum.....
<img width="782" alt="image" src="https://github.com/Secret-Lab-Productions/heizoel24-mex/assets/5463753/cb82e65a-e8b3-451e-8764-2b9c56d2d54d">

**Reply by ltspicer** [reply]
Ich arbeite zur Zeit an einem ioBroker Adapter. Der ist bereits in den Startlöchern.
Da ich kein HA benutze, werd ich nicht an einer Lösung für diesen arbeiten. Sorry.

**Reply by Marcwa19197** [reply]
Aber dann wird dein MEXX ja trotzdem die Daten zuerst an einen öffentlichen Cloud-API-Server "rausschicken" oder geht das dann rein Lokal?

**Reply by ltspicer** [reply]
Nein. Wird immer von der HeizOel24 Cloud lesen.
Lokal hab ich nicht geschafft. Hab ja auch keine Ahnung, wie das Protokoll des MEX aussieht.

**Reply by Marcwa19197** [reply]
Okay, verständlich. Ich würde einfach mal hier im GitHub Projekt (hoffe das ist der richtige Platz dafür) eine zusätzliche Diskussion dafür öffnen. Vielleicht hat ja noch jemand Bedarf dafür und kann unterstützen. Meine Python Kenntnisse sind relativ eingestaubt. Ich bin kein "richtiger" Entwickler, eher ein Sysadmin.

Edit: Ich habe mal eine neue Diskussion geöffnet: https://github.com/Secret-Lab-Productions/heizoel24-mex/discussions/3

**Reply by ltspicer** [reply]
Ist jemand hier, der 2 MEX Geräte hat?
Das konnte ich nämlich bis jetzt nicht testen.

**Reply by Marcwa19197** [reply]
Zwei MEX habe ich leider nicht, aber eine andere Frage:
- Wieso wurde der Verzögerungstimer (bis 59 min) eingebaut? Gibt es eine API Limitierung, dass man ansonsten ggf. blockiert wird, wenn man die API-Server zu oft anfragt?
Habe das Script soeben geupdated und meine HomeAssistant Konfig angepasst (Die ID "1" im MQTT-Topicpfad hinzugefügt). Funktioniert einwandfrei! :)

**Reply by ltspicer** [reply]
Wenn viele dieses Script benutzen, würden alle immer um 0 */3 die Daten abfragen.
Dann könnte ich mir vorstellen, das heizoel24 plötzlich sperrt.
So wird der Server nicht immer zur selben Minute gehämmert.

**Reply by Marcwa19197** [reply]
Macht Sinn ja, konnte auch beobachten, dass der MEX die Daten nicht immer zur exakten Uhrzeit sendet (bei mir meist Nachts so zwischen 02-04 Uhr)

**Comment by Olly77Wolly** [comment]
MoinMoin, ich hatte die Daten einmal in das custom_components Verzeichnis importiert. Leider bekomme ich nach einem Neustart keine neue Geräte Integration mit dem Namen MEX oder Heizoel24 angezeigt.

**Comment by Olly77Wolly** [comment]
Muss ich evtl. doch Ecofrog als Integration verwenden? Mir werden beim hinzufügen leider nur vier freie Felder zum ausfüllen angezeigt aber ohne Bezeichnung. Wo bekomme ich diese UUID her?

**Reply by ltspicer** [reply]
Bei meinem Python Script wird die UUID automatisch gezogen.
Sonst kannst mit Charles Proxy und dessen ssl Library den Datenverkehr der App auslesen. Da siehst unter Anderem auch die UUID.

**Comment by Olly77Wolly** [comment]
Wird bei Euch den die Mex Integration beim hinzufügen erkannt oder muss ich die Ecofrog Integration dafür nutzen? Müssen auf dem custom_components Verzeichnis noch Rechte vergeben werden? Ich bin leider noch nicht so lange mit dem HA unterwegs und auch kein Programmierer :( .

**Reply by MonsterD1** [reply]
I used the initial work version to see the mex integration: [link](https://github.com/Secret-Lab-Productions/heizoel24-mex/tree/initial_work)

**Comment by MonsterD1** [comment]
getting the error:
`ERROR (SyncWorker_59) [custom_components.heizoel24-mex.config_flow] {'Username': 'xxx@xxxx.dexx', 'Password': 'xxxxx-xxxxxx-xxxxxx'}
2024-03-04 10:28:44.591 ERROR (SyncWorker_59) [custom_components.heizoel24-mex.config_flow] Authenticating: {"SessionId":"4daf7f19-2df6-42b4-8a07-e4df959d7ec8","UsernameDoesNotExist":false,"WrongPassword":false,"HasMexEnabled":true,"ResultCode":0,"ResultMessage":null}`

I am sure the user / password is correct... any suggestion on how to solve it?
